### PR TITLE
pad_list_data_collate

### PR DESCRIPTION
**Pull request by rijobro**
A collator, which prior to performing the running collator, pads the images to make sure that sizes match for a given batch.

This is done by calculating the maximum size in each dimension, and then padding all images in the batch to that size.

This introduced a slight circular dependency, because the spatialpadd is now used inside the data level. Solving it required all the seemingly unrelated changes to other files.

### CODE REVIEW

File: tests/test_pad_collation.py [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+from typing import List, Tuple
+
+import numpy as np
+from parameterized import parameterized
+
+from monai.data.utils import pad_list_data_collate
+from monai.transforms import RandRotate90d, RandRotated, RandSpatialCropd, RandZoomd
+from monai.utils import set_determinism
+
+set_determinism(seed=0)
```
Comment (wyli) [inline tests/test_pad_collation.py:22]: is it possible to move this into `setUp`? e.g. https://github.com/Project-MONAI/MONAI/blob/db831ae096bec76371bb8c2e3db226d947a5ac06/tests/test_bounding_rect.py#L28-L32

File: monai/data/utils.py [hunk 2]
```diff
@@ -240,7 +241,57 @@ def list_data_collate(batch: Sequence):
...
+        if "stack expects each tensor to be equal size" in re_str:
+            re_str += (
+                "\nMONAI hint: if your transforms intentionally create images of different shapes, creating your "
+                + "`DataLoader` with `collate_fn=pad_list_data_collate` might solve this problem (check its "
+                + "documentation)."
+            )
+        raise RuntimeError(re_str)
+
+
+def pad_list_data_collate(batch: Sequence):
```
Comment (wyli) [inline monai/data/utils.py:257]: this mainly support `batch` with a list of dictionary, shall we support list of list as well? the underlying `default_collate` is quite powerful: https://github.com/pytorch/pytorch/blob/master/torch/utils/data/_utils/collate.py#L43

**Review by wyli — COMMENTED** [review]
thanks! maybe we have some issues with the overall utility module arrangements...please see some comments inline

File: tests/test_pad_collation.py (hunk 1 above)
Comment (rijobro) [reply tests/test_pad_collation.py]: Done!

File: monai/data/utils.py (hunk 2 above)
Comment (rijobro) [reply monai/data/utils.py:257]: Although this is doable, I can't immediately see the use case. Currently I only want to decollate for inverse transformations which means the format is list of dictionaries. Would you accept this?

```python
if isinstance(batch[0], list):
    raise NotImplementedError("decollation has only currently been implemented for list of dictionaries. Please raise a PR if you feel this functionality would be useful")
```

Comment (wyli) [reply monai/data/utils.py:257]: sure, sounds great

Comment (rijobro) [reply monai/data/utils.py:257]: In the end it was pretty simple to allow for list of lists/tuples, so I added that too.

Comment (wyli) [reply monai/data/utils.py:257]: thanks, forgot to say that we should support different padding mode here...

Comment (rijobro) [reply monai/data/utils.py]: Good point, I've just added padding options

**Review by wyli — APPROVED** [review]
thanks, this looks nice

**Pull request merged**